url47e6aca666_pocorall_automaton_tgz-pJ8-net_pocorall_automaton_DatatypesJ8/src/net/pocorall/automaton/BasicAutomataFactory.java

```diff
@@ -0,0 +1,502 @@
+	}
+
+	/**
+	 * Constructs automaton that accept strings representing decimal numbers
+	 * that can be written with at most the given number of digits in the fraction part.
+	 * Surrounding whitespace is permitted.
+	 *
+	 * @param i max number of necessary fraction digits
+	 */
+	public static DefaultAutomaton makeFractionDigits(int i) {
+		return ((new RegExp("[ \t\n\r]*[-+]?[0-9]+(\\.[0-9]{0," + i + "}0*)?[ \t\n\r]*")).toAutomaton()).minimize();
+	}
+
+	/**
+	 * Constructs automaton that accept strings representing the given integer.
+	 * Surrounding whitespace is permitted.
+	 *
+	 * @param value string representation of integer
+	 */
+	public static DefaultAutomaton makeIntegerValue(String value) {
+		boolean minus = false;
+		int i = 0;
+		while (i < value.length()) {
+			char c = value.charAt(i);
+			if (c == '-')
+				minus = true;
+			if (c >= '1' && c <= '9')
+				break;
+			i++;
+		}
+		StringBuilder b = new StringBuilder();
+		b.append(value.substring(i));
+		if (b.length() == 0)
+			b.append("0");
+		DefaultAutomaton s;
+		if (minus)
+			s = makeChar('-');
+		else
+			s = makeChar('+').optional();
+		DefaultAutomaton ws = Datatypes.getWhitespaceAutomaton();
+		return (ws.concatenate(s.concatenate(makeChar('0').repeat()).concatenate(makeString(b.toString()))).concatenate(ws)).minimize();
+	}
+
+	/**
+	 * Constructs automaton that accept strings representing the given decimal number.
+	 * Surrounding whitespace is permitted.
+	 *
+	 * @param value string representation of decimal number
+	 */
+	public static DefaultAutomaton makeDecimalValue(String value) {
+		boolean minus = false;
+		int i = 0;
+		while (i < value.length()) {
+			char c = value.charAt(i);
+			if (c == '-')
+				minus = true;
+			if ((c >= '1' && c <= '9') || c == '.')
+				break;
+			i++;
+		}
+		StringBuilder b1 = new StringBuilder();
+		StringBuilder b2 = new StringBuilder();
+		int p = value.indexOf('.', i);
+		if (p == -1)
+			b1.append(value.substring(i));
+		else {
+			b1.append(value.substring(i, p));
+			i = value.length() - 1;
+			while (i > p) {
+				char c = value.charAt(i);
+				if (c >= '1' && c <= '9')
+					break;
+				i--;
+			}
+			b2.append(value.substring(p + 1, i + 1));
+		}
+		if (b1.length() == 0)
+			b1.append("0");
+		DefaultAutomaton s;
+		if (minus)
+			s = makeChar('-');
+		else
+			s = makeChar('+').optional();
+		DefaultAutomaton d;
+		if (b2.length() == 0)
+			d = makeChar('.').concatenate(repeat(makeChar('0'), 1)).optional();
+		else
+			d = makeChar('.').concatenate(makeString(b2.toString())).concatenate(makeChar('0').repeat());
+		DefaultAutomaton ws = Datatypes.getWhitespaceAutomaton();
+		return (ws.concatenate(s.concatenate(makeChar('0').repeat()).concatenate(makeString(b1.toString())).concatenate(d)).concatenate(ws)).minimize();
+	}
+
+	/**
+	 * Constructs deterministic automaton that matches strings that contain the given substring.
+	 */
+	public static DefaultAutomaton makeStringMatcher(String s) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 12. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.